oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/OpenIdConfigurationJwtDecoderProvider.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.jwt;
+
+import com.nimbusds.oauth2.sdk.ParseException;
+import com.nimbusds.openid.connect.sdk.op.OIDCProviderMetadata;
+
+import org.springframework.web.client.RestTemplate;
+
+/**
+ * Allows creating a {@link JwtDecoder} from an
+ * <a href="https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig">OpenID Provider Configuration</a>.
+ *
+ * @author Josh Cummings
+ * @since 5.1
+ */
+public final class OpenIdConfigurationJwtDecoderProvider {
```

Comment:
This class needs to be in the `config` module just like `OidcConfigurationProvider` is.

Also, the name of this class restricts it to `JwtDecoder` configuration only. I think we need to rename it to be more generic so it makes sense when we need other provider configuration and associated static factory methods.

The common naming convention is to prefix with `Oidc` not `OpenId`

Reply:
> Can you clarify why the existing `OidcConfigurationProvider` is in the config module

This is considered Java Config, just like `OAuth2ResourceServerConfigurer` and `OAuth2ClientConfigurer`, hence it belonging in `config` module

> the only generic name that makes sense to me is `OidcConfigurationProvider`

Yeah same for me. Although we have this [ticket](https://github.com/spring-projects/spring-security/issues/5584) to rename to `OidcProviderConfiguration`

>  What are your thoughts on two classes of the same name, in different packages

I was thinking the same and initially it makes sense but at the same time having users (or Boot) to fully qualify the class name when both are used (Client and Resource Server) would be a pain. I'm not sure what other name we can use at the moment.

> Or what rules do we have for moving the existing OidcConfigurationProvider into a more generic package like o.s.s.config.oauth2.oidc.OidcConfigurationProvider which could then house both methods?

Housing both methods in `OidcConfigurationProvider` would require both dependencies - client and resource server - which is not good especially if the user is only using client for example. 

However, I believe @rwinch did suggest this strategy (housing in one config class) when we spoke? Maybe he can clarify this further as I'm not sure if I understood him correctly.

> I we can combine the two, then I might like something named OidcDiscoveryConfigurationProvider

I think we should use the name `OidcProviderConfiguration` to align with the spec. Or at least have that in the class name.

I do like your implementation approach on caching the `OIDCProviderMetadata` as per

> public static OidcConfigurationProvider issuer(String issuer)

Reply:
> I think we should use the name OidcProviderConfiguration to align with the spec. Or at least have that in the class name.

My bad, I had misread the spec when I saw "Discovery" in the title [1], forgetting that we are just focused on the configuration section.

[1] - https://openid.net/specs/openid-connect-discovery-1_0.html#IssuerDiscovery